File: ABC_from101to200/ABC189/e.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;
#define rep(i, n) for (int i = 0; i < n; i++)
using vll = vector<ll>;
using vvll = vector<vll>;

vvll multiple33(vvll a, vvll b)
{
    vvll ret(3, vll(3));
    rep(i, 3) rep(j, 3) rep(k, 3) ret[i][j] += a[i][k] * b[k][j];
    return ret;
}

vll multiple31(vvll a, vll b)
{
    vll ret(3);
    rep(i, 3) rep(j, 3) ret[i] += a[i][j] * b[j];
    return ret;
}

vvll op3(ll p)
{
    return {{-1, 0, 2 * p}, {0, 1, 0}, {0, 0, 1}};
}

vvll op4(ll p)
{
    return {{1, 0, 0}, {0, -1, 2 * p}, {0, 0, 1}};
}

int main()
{
    vvll op1 = {
        {0, 1, 0},
        {-1, 0, 0},
        {0, 0, 1}};
    vvll op2 = {
        {0, -1, 0},
        {1, 0, 0},
        {0, 0, 1}};
    int n, m, q;
    cin >> n;
    vvll x(n, {0,0,1});
    rep(i, n) cin >> x[i][0] >> x[i][1];
    cin >> m;
    vector<vvll> ops(m + 1);
    ops[0] = {{1, 0, 0}, {0, 1, 0}, {0, 0, 1}};
    rep(i, m)
    {
        int op;
        ll p = 0;
        cin >> op;
        if (op == 3 || op == 4)
            cin >> p;
        switch (op)
        {
        case 1:
            ops[i + 1] = multiple33(op1, ops[i]);
            break;

        case 2:
            ops[i + 1] = multiple33(op2, ops[i]);
            break;

        case 3:
            ops[i + 1] = multiple33(op3(p), ops[i]);
            break;

        case 4:
            ops[i + 1] = multiple33(op4(p), ops[i]);
            break;

        default:
            break;
        }
    }
    cin >> q;
    rep(i, q)
    {
        int a, b;
        cin >> a >> b;
        vll ans = multiple31(ops[a], x[b - 1]);
        cout << ans[0] << " " << ans[1] << endl;
    }
}

File: ABC_from101to200/ABC189/e.py
import numpy as np


def op1():
    return np.array([[0, 1, 0], [-1, 0, 0], [0, 0, 1]], dtype=np.int64)


def op2():
    return np.array([[0, -1, 0], [1, 0, 0], [0, 0, 1]], dtype=np.int64)


def op3(p):
    return np.array([[-1, 0, 2 * p], [0, 1, 0], [0, 0, 1]], dtype=np.int64)


def op4(p):
    return np.array([[1, 0, 0], [0, -1, 2*p], [0, 0, 1]], dtype=np.int64)


def operation(o, p=0):
    if o == 1:
        return op1()
    elif o == 2:
        return op2()
    elif o == 3:
        return op3(p)
    else:
        return op4(p)


n = int(input())
orign = []
for i in range(n):
    x, y = map(int, input().split())
    orign.append(np.array([x, y, 1]))
m = int(input())
cum_op = [np.array([[1, 0, 0], [0, 1, 0], [0, 0, 1]], dtype=np.int64)]
for i in range(m):
    der = list(map(int, input().split()))
    ori = cum_op[-1]
    if der[0] == 1 or der[0] == 2:
        cum_op.append(operation(der[0])@ori)
    else:
        cum_op.append(operation(der[0],der[1])@ori)
q = int(input())
ans=[]
for i in range(q):
    a, b = map(int, input().split())
    mul = np.dot(cum_op[a], orign[b - 1])
    ans.append((mul[0], mul[1]))

for v in ans:
    print(v[0], v[1])
